core/state_transition.go

```diff
@@ -374,7 +374,10 @@ func (st *StateTransition) TransitionDb() (*ExecutionResult, error) {
 	st.gas -= gas
 
 	// Check clause 6
-	if msg.Value().Sign() > 0 && !st.evm.Context.CanTransfer(st.state, msg.From(), msg.Value()) {
+	// Note: If this is an L1MessageTx, we will not return a top-level ErrInsufficientFundsForTransfer.
+	// Instead, we return ErrInsufficientBalance from within st.evm.Call. This means that such transactions
+	// will revert but they can be included in a valid block. This is necessary for supporting enforced txs.
+	if msg.Value().Sign() > 0 && !st.evm.Context.CanTransfer(st.state, msg.From(), msg.Value()) && !msg.IsL1MessageTx() {
```

Comment:
Note about backward compatibility:

While this is a concensus rule, this case is impossible to trigger now: `EnforcedTxGateway` is paused so users cannot construct arbitrary L1 messages.